### ISSUE TITLE
API feedback

### ISSUE THREAD

**Issue by MarkPflug** [AllocZero/EfficientDynamoDb#100]
Here are a few notes on the API that I've noticed reading through the documentation. I don't know how open you are to changing the API surface area at this point. Given that it is still pre-1.0, now would be the best time to consider these.

### EqualsTo

In this example: `filter.On(x => x.FirstName).EqualsTo(firstNameValue)`

"EqualsTo" sounds wrong to me. I think it should be one of: "EqualTo", "IsEqualTo", or "Equals" (in order of my preference). I think if "IsEqualTo" is chosen, then all the other operators should be updated to match, ie "IsGreaterThan". "Equals" might be bad because it collides with `object.Equals`.

### Logical operators API.

I think these operator should be changed from the bitwise operators ("&", "|") to the logical operators ("&&", "||").

I'll update this issue with any other things that I notice.

**Comment by MarkPflug** [comment]
### Arithmetic operations

This example from update expressions:
```csharp
ddbContext.UpdateItem<EntityClass>()
    .On(x => x.Age).AssignSum(x => x.Age, 1) // increment age
    .On(x => x.CacheTtl).AssignSubtraction(x => x.CacheTtl, 10) // reduce cache TTL by 10 (seconds)
    .ExecuteAsync();
```

The `AssignSum` and `AssignSubtraction` names sound strange to me. Perhaps `Increment` and `Decrement` instead? Is it possible to not need to repeat the target property?

I would ideally like to write:
```csharp
ddbContext.UpdateItem<EntityClass>()
    .On(x => x.Age).Increment() // increment age by 1
    .On(x => x.CacheTtl).Decrement(10) // reduce cache TTL by 10 (seconds)
    .ExecuteAsync();
```

This later example:
```csharp
ddbContext.UpdateItem<EntityClass>()
    .On(x => x.Age).AssignSum(x => x.FirstProperty, x => x.SecondProperty)
    .ExecuteAsync();
```

It would be nice to be able to write:
```csharp
ddbContext.UpdateItem<EntityClass>()
    .On(x => x.Age).Assign(x => x.FirstProperty + x.SecondProperty)
    .ExecuteAsync();
```

This would require more complex analysis/transformation of the expression, but the API would feel more intuitive in my opinion.

**Comment by firenero** [comment]
> "EqualsTo" sounds wrong to me. I think it should be one of: "EqualTo", "IsEqualTo", or "Equals" (in order of my preference). I think if "IsEqualTo" is chosen, then all the other operators should be updated to match, ie "IsGreaterThan". "Equals" might be bad because it collides with object.Equals.

Absolutely agree, I will rename it to `EqualTo`

> I think these operator should be changed from the bitwise operators ("&", "|") to the logical operators ("&&", "||").

Overloading logical operators is not allowed in C#. So the best way we found was overloading bitwise operators.

**Comment by firenero** [comment]
> The AssignSum and AssignSubtraction names sound strange to me. Perhaps Increment and Decrement instead? Is it possible to not need to repeat the target property?

Yep, `Increment()` and `Decrement()` overloads sounds great, and it's straightforward to add them. I'll do it soon. 
`AssignSum` and `AssignSubtraction` are not the best names for sure, but I couldn't think about something better back then. I'm very open to name suggestions here.

```csharp
Assign(x => x.FirstProperty + x.SecondProperty)
```

This example looks cool on paper, but it has many hidden issues that make implementation complicated or even not worth the effort. 
1. DDB only supports `+` and `-` operations, but expressions would allow you to use any other operators like `/`, `*`, `%`, etc. And these errors won't be visible at compile time but constantly crashing at runtime. 
2. It becomes tough to support DDB `if_not_exists()` function with such API., And even if we implement it, the actual API may become ugly.

Overall, I feel like this API is hard to add and even harder to fix/update if new DDB functionality is released or some bugs are found.

**Comment by MarkPflug** [comment]
>This example looks cool on paper, but it has many hidden issues that make implementation complicated

Understood. I was basically envisioning a Linq to dynamo style API. Implementing a query transformation can be confusing, and indeed has the issue that the C# language tends to be more expressive than the target query language. You can validate the expression at compile via roslyn analyzers, which could verify that only + or - are used in this context, for example, but that is additional work for sure.

**Comment by lezzi** [comment]
@firenero is on vacation this week so we probably will continue improving the API next week.

**Comment by firenero** [comment]
@MarkPflug The following API updates are available in `0.9.2-alpha.0.4`:

* `Increment()/Decrement()` methods
* `EqualsTo` -> `EqualTo` renaming (including other methods that contain `Equal` in name)

LINQ-like assign syntax is not something we want to add at this point. It seems that it creates more problems than it tries to solve now. But we might want to revisit it in the future.

I'm going to close this issue. Feel free to reopen it if you have something to add. However, for new API suggestions, I'd recommend creating a separate issue. Just to keep issues size manageable.
